arches_for_science/trigger_functions.py

```diff
@@ -0,0 +1,184 @@
+INTO fn_config
+FROM functions_x_graphs
+WHERE
+    functionid = '00b2d15a-fda0-4578-b79a-784e4138664b'
+    AND functions_x_graphs.config IS NOT NULL
+    AND graph = graphid;
+
+IF FOUND THEN
+    SELECT JSONB_OBJECT('{}') INTO localized_result_all;
+    SELECT JSONB_OBJECT('{}') INTO localized_result_name_only;
+    SELECT fn_config -> 'descriptor_types' -> 'name' -> 'string_template' INTO name_template;
+    SELECT fn_config -> 'descriptor_types' -> 'map_popup' -> 'string_template' INTO map_popup_template;
+    SELECT fn_config -> 'descriptor_types' -> 'description' -> 'string_template' INTO description_template;
+
+    SELECT ARRAY[name_template, map_popup_template, description_template] INTO all_templates;
+    SELECT 0 INTO loop_index;
+    FOR this_template IN SELECT UNNEST(all_templates)
+    LOOP
+        SELECT loop_index + 1 INTO loop_index;
+        SELECT
+        CASE
+            WHEN loop_index = 1 THEN 'name'
+            WHEN loop_index = 2 THEN 'map_popup'
+            ELSE 'description'
+        END
+        INTO descriptor_key;
+
+        -- Resolve node values to localized strings
+        FOR alias_with_separators IN SELECT UNNEST(REGEXP_MATCHES(this_template, '<\w+>', 'g'))
+        LOOP
+            SELECT TRIM(BOTH '<>' FROM alias_with_separators) INTO node_alias;
+
+            SELECT nodeid, nodegroupid INTO node_for_alias, nodegroup_for_alias
+                FROM nodes
+                WHERE alias = node_alias
+                AND graphid = graph;
+
+            SELECT tiledata ->> node_for_alias::text INTO localized_string_node_value
+                FROM tiles
+                WHERE resourceinstanceid = resourceid
+                AND nodegroupid = nodegroup_for_alias
+                AND sortorder = 0
+                LIMIT 1;
+
+            -- Replace template with localized string
+            FOR lang, inner18n_obj_for_lang IN SELECT * FROM jsonb_each(localized_string_node_value) 
+            LOOP
+                -- Initialize language key if missing
+                IF localized_result_all -> lang IS NULL THEN
+                    SELECT jsonb_set(
+                        localized_result_all,
+                        ARRAY[lang],
+                        JSONB_OBJECT('{}')
+                    ) INTO localized_result_all;
+                END IF;
+
+                -- Retrieve the current working value, or start from the template
+                SELECT
+                CASE
+                    WHEN localized_result_all -> lang -> descriptor_key IS NULL
+                        THEN this_template
+                    ELSE TRIM((localized_result_all -> lang -> descriptor_key)::text, '\"')
+                END
+                INTO working_string;
+
+                SELECT REPLACE(
+                    working_string,
+                    alias_with_separators,
+                    TRIM((inner18n_obj_for_lang -> 'value')::text, '"')
+                )
+                INTO resolved_node_value_for_lang;
```

Comment:
What should the fallback value be when there is no value for the given node? Leave `Prefix <statement_content> Suffix`, which is at least somewhat informative (and what happens now), or just `Prefix Suffix`?

When it's just that a language is missing on the node vlalue being worked on, currently the fallback is replace the alias with nothing/empty string (like the latter option just mentioned). At the moment this is just a happy accident of the way languages are stored as `{"de": "value": "" ...}` instead of `{"de": "value": null, ...}`.

Reply:
Updated in [13f025d](https://github.com/archesproject/arches-for-science/pull/1452/commits/13f025da92ae955a169a138f3267e7cf772d96f3) to fall back to ` -- ` when a value for a node is missing.

Missing languages still just fall back to whitespace.